### PR TITLE
Dev v7 health check

### PR DESCRIPTION
**Pull request by nul800sebastiaan**
[U4-3805 Add "Health check" list dashboard to the core](http://issues.umbraco.org/issue/U4-3805)

I've turned this into a PR so we can more easily follow and discuss the progress of this branch, it will be pulled in once we have some of the mechanics working:
- A dashboard with checks divided in groups
- "Check"/"Recheck" and "Fix" (where possible) buttons on each check in the dashboard
- At least a known working external healtcheck (so a dll to drop in that isn't provided by the core) to prove extensibility works

I think that's the MVP (minimal viable product) and then we can iterate on it and implement new ideas that might come up.

### CODE REVIEW

**Comment by nul800sebastiaan** [comment]
@dampee You're correct! Fixed the problems you mentioned. I think `Umbraco.Web.HealthCheck` is a fine namespace. I don't think we'll implement a large amount of checks in the core, rather I think the community should come up with a good bundle of them that can be dropped in easily depending on the needs for each site.

That said, I'd also love to come up with a list of essential checks that we think should be in the core!

On the top of my personal list is security checks, like @kgiszewski started with [in his forum post](https://our.umbraco.org/forum/umbraco-7/developing-umbraco-7-packages/70277-security-essentials-proposal).

I'm off to build a fun "[HaveIBeenPwnd](https://haveibeenpwned.com/)" check now, to make sure plugin checks can work.

**Comment by nul800sebastiaan** [comment]
Super rudementary dashboard added now.
image

**Comment by nul800sebastiaan** [comment]
Success!
image

So this is an external dll with a "HaveIBeenPwnd" HealtCheck in it. It works both as a dll and as a .cs file to drop into App_Code.

https://gist.github.com/nul800sebastiaan/1d3cc51deedae8dbd26c33e597edd3db

@Shazwazza I assume the `Actions` were meant to be a list of ways to fix any errors found? So in this case I've referred to an action `sendNotificationMails` which should trigger the `SendBreachNotifications` method.

**Comment by bjarnef** [comment]
@nul800sebastiaan there is a typo here: https://gist.github.com/nul800sebastiaan/1d3cc51deedae8dbd26c33e597edd3db#file-haveibeenpwndcheck-cs-L19

https://haveibeenpwnd.com should be https://haveibeenpwned.com (missing an "e") :sunglasses:

**Comment by abjerner** [comment]
@nul800sebastiaan Are there some deep thoughts behind the current dashboard (and the JSON returned by the `GetAllHealthChecks` API method, or is it just a quick example? Currently it will just list the status of each health status, but without any information regarding the health check or group?

From my example package from the Our thread a while back, the dashboard could look something like this:

image

A benefit here is also that each check isn't executed by default, so time consuming checks wouldn't be executed when the user visits the dashboard.

@Shazwazza I tried asking asking on the issue tracker why the health check code is using an attribute for discovering classes. I'm not sure I got an answer for this, so what is the reason for this approach? Personally I would find the approach similar to how you add an `IContentFinder` much easier and would find it more flexible.

**Comment by nul800sebastiaan** [comment]
Yes this is all just prototype/proof of concept level. I just wanted to make sure we could get something that works for now and people could help out with! The `GetAllHealthChecks` actually EXECUTES all of the checks currently as well which is not good, but I just wanted to make sure that all the moving parts basically worked so the community could pick up on the actual API and the interface (looks great, your example!) implemented.

**Comment by abjerner** [comment]
Oki. Just making sure. I have pushed some minor changes for this to my own fork: https://github.com/abjerner/Umbraco-CMS/commit/c828f120ecad98902bdbf5a2b5dad21ae3149be3

How can I add the changes to this pull request? I'm not used to more people working on the same pull request, so I only know how to make a new pull request :(

**Comment by nul800sebastiaan** [comment]
@abjerner Easy enough: send us a PR and instead of targeting the `dev-v7` branch, you target the  `dev-v7-HealthCheck` branch!

**Comment by nul800sebastiaan** [comment]
@bjarnef That was not really the point, it was just an example ;-) However, it's now fixed and the group is set to `Security` instead now.

So with @abjerner 's great updates it's looking like this now:
image

@abjerner I guess we need a description for the fix action as well? Right now it's not very friendly and just a method name. And maybe the fix action should be a button with a tooltip or something?

**Comment by abjerner** [comment]
@nul800sebastiaan Awesome. I still have a little trouble with the UI. If we are to display more information about each action, we may need to show them in a nested table or perhaps in a dialog. I'm not sure yet what is the best way to do this, but I'll try to play a bit around with the UI and see what works.

And a friendly name and a description would indeed be useful for the actions :smiley:

**Comment by nul800sebastiaan** [comment]
Feels like it's crammed to the right, maybe just list the results under the check.
It also seems like we should have a "Run all checks in this group" button. 

I am also working on an HTTPS-healthcheck and am noticing that I want to do multiple checks in one, and making the severity different depending on those, example:

* Site doesn't run on HTTPS: error
* Site runs on HTTPS and `umbracoUseSSL` is false: warning
* Site runs on HTTPS and `umbracoUseSSL` is true: success

I want to write multiple result messages for that, to list the results for both checks (there might be more). The messages might also get quite long, HTTPS configuration requires some explanation. However, that can be linked to documentation too.

The checks depend on each other, and each of result messages should also have a result icon in front of them.

**Comment by nul800sebastiaan** [comment]
Ps. Shannon is on holiday, I'll nudge him when he's back. :)

**Comment by nul800sebastiaan** [comment]
Ah, I jJust realized the XmlDataIntegrityCheck is doing what I want, multiple checks in one!

**Comment by Shazwazza** [comment]
@abjerner To follow up on your questions - Health checks are *not* discovered by attributes this is why there is a base class `HealthCheck`, you can see the discovery here: https://github.com/umbraco/Umbraco-CMS/pull/1220/files#diff-99785536d6d3ec0f5e535d4e7aac4f61R539 

Attributes are used to describe the metadata, that is all. 

Also note that we are currently doing auto-discovery. We could turn this off and force developers to manually register on startup but that isn't exactly making peoples lives very easy. It would mean that for any additional health check package a dev would need to manually add it. If you want full control over the ordering or the specific checks listed, you can still do this (I mentioned this to you in the original issue: http://issues.umbraco.org/issue/U4-3805#comment=67-22929). Currently because this is a lazy resolver, I might need to add some functionality to be able to clear the laziness during startup and allow a developer to replace it with a custom list, should be easy though.

**Comment by nul800sebastiaan** [comment]
Work in progress on a HTTPS has just been added.
Noticed that we need something like a spinner while a test is in progress, in this case I'm trying to ping a URL in one of the tests and that might take a few seconds to return a result.

I also think it would be nice if the Succes/Warn/Error icons were shown in their correct colors that we use for the rest of the backoffice.

image

**Comment by abjerner** [comment]
@Shazwazza Oki. Perhaps I'm not sure how that part works "under the hood" in Umbraco. I personally (as a developer) wouldn't mind having to add checks manually. I guess we can skip this part for now, and then come back to it later (if necessary).

@nul800sebastiaan I had some similar thoughts regarding the UI. A loader is IMO a must since an action can take more than just a few seconds. A colored icon would also help identifying potential issues, so I'll try and play around with that in the UI.

**Comment by abjerner** [comment]
A little progress for the UI. Each check now also have a proper loader (not visible in the screenshot). I should be able to create another pull request tomorrow with updates to the UI and a bit of the underlying API as well ;)

image

**Comment by nul800sebastiaan** [comment]
Looks much nicer! I realize it's not always possible to do an automatic fix for some of the issues. In this case some text and a link to more information is probably the best way to go. For the things we CAN fix automatically, it would be nice to add an explanation as to what is going to happen when we apply the fix (for example: "this fixe updates the web.config file, so proceed with caution, your application pool will recycle and you might experience some downtime").
And the name of the  And then instead of a link to the fix method, that should just be a "Fix" button. I don't think "Fix all" is going to be a good experience, maybe if there's loads to fix? I think it will promote carelessness and people not reading before they do something that's potentially going to cause downtime.

**Comment by Shazwazza** [comment]
Also keep in mind that this class contains other properties that haven't been taken into account yet:

https://github.com/umbraco/Umbraco-CMS/pull/1220/files#diff-27634fb7c6067506302c33dba50479f3R12

This should allow a developer to specify a custom angular view to render as the result - so they could actually show basically whatever they want. This may not be required or might be too flexible, but if we require it we should pass in a context of some sort to that view so it can know what check ID it is being run for, and any other bits of info it might need.

Also the enumerable result: `public IEnumerable<HealthCheckAction> Actions { get; set; }` could depict what buttons are displayed (i.e. could contain button text)

**Comment by abjerner** [comment]
I have created a new pull request with the latest changes: https://github.com/umbraco/Umbraco-CMS/pull/1237

Regarding the part for executing actions, I'm not totally sure about the code yet. In @Shazwazza's skeleton code, there is a `HealthCheckStatus ExecuteAction(HealthCheckAction action)` method in the API controller. Is there a reason for providing the entire model to the API method, or was that just part of the initial skeleton? At least I'm not sure how to call the method that way, so I have instead added my own `HealthCheckStatus ExecuteAction(string checkId, string actionAlias)`. But my implementation is perhaps a bit "questionable" (see [this line](https://github.com/umbraco/Umbraco-CMS/blob/7d3fd9ce37a9115b8b20eea28a991badf59ec4e6/src/Umbraco.Web/HealthCheck/HealthCheckController.cs#L76)). Also, is there a reason for having both an alias and an ID/GUID? Having both seems a bit redundant - at least now since I've added a name as well (which is shown in the UI as the link text for the action).

@nul800sebastiaan I think I've misunderstood you about the buttons, so I have removed the "Fix all" button again.

As part of my changes for actions, I have also added a description, which is shown in the tooltip of the action link. If this is to inform people that they are about to do something dangerous, it's probably hidden to much away just being a tooltip, but I haven't yet come up with the right way to show in the UI otherwise.

@Shazwazza I hadn't seen the part with the view for statuses, but it seems like a good idea. I'll try to add support for that later. I think if well use `ng-transclude`, the view should have access to `check` and `status` of the parent scopes so developers will have access to it in their own views/controllers.

**Comment by nul800sebastiaan** [comment]
Hey @abjerner I had these changes as good as ready for a week, but with some travelling I could never get them done. I will review your PR shortly but I'm afraid there's a bit of an overlap so some adjustments on your end will be needed. But I have some concerns too so don't start yet before the review is in! :smile: 

In this commit:
- Updates to add some color and change the layout a bit
- Adds a "Check all" button for groups (TODO: "Check all checks in all groups")
- Adds name/description for a fix button so that it can be customized if the label "Rectify" isn't enough
- Adds a fix to change umbracoUseSSL from false to true (or add it)

It looks a little like this now (we'll get our colleague Simon to improve the UI soon).

image

This uses the returned actions for each check, whereas I had a look at your code and I couldn't quite figure out when a "rectify" link would appear (I didn't look very hard though).
If you don't give the action a name/description it will just show a "Rectify" button.

I have not implemented a UI for the "rectify" results yet.

**Comment by nul800sebastiaan** [comment]
Feedback after fixing something is now also shown in the UI.

**Comment by nul800sebastiaan** [comment]
Okay, I have gone through your code @abjerner and understand what it's doing now. I've not pulled in your changes but copied most of it and adapted it to the "Umbraco style", great stuff! :+1: 

Some changes I made:

- Removed `OneIsAFlock` and just return a `new []` 
- Refactored `string Value` into `IEnumerable<AcceptableConfiguration> Values` - I noticed that `MacroErrors` in `umbracoSettings.config` could have either `inline` or `silent` as a valid value and wanted to check for both. Same goes for `CustomErrors`, the `mode` can be either `On` or `RemoteOnly`, which are perfectly fine values
- Added `CurrentValue` so we can display that in the check
- Fixed the `TraceCheck` check, the Rectify button would set the value to `RemoteOnly` (copy/paste error, I'm sure)
- Fixed `AbstractConfigCheck.Rectify()`, it would try to set `node.Value` but in the case of `MacroErrors` it needs to set `node.InnerText`

Some of the new checks look like this:

image

**Comment by nul800sebastiaan** [comment]
One thought I had while hardcoding all these strings: we should be able to localize the checks. However, I do believe it would be very difficult to make the attributes translatable, do you have any thoughts on how to do this best @Shazwazza ?

**Comment by abjerner** [comment]
@nul800sebastiaan Awesome. I haven't really had the time to follow up on your comments. But I'll probably have another look at it over the weekend ;)

I'm sorry if I didn't make it clear, but some of the code was "temp" code since I wasn't 100% sure about the structure, and thus planned to fix it and make it "prettier" later. Also, the checks that I added are more or less copied from the the old uGoLive project. As far as I remember, there are still some other checks in that package that also could be useful for the health check dashboard.

**Comment by nul800sebastiaan** [comment]
@abjerner No worries, I saw that you did indeed copy some of uGoLive when I was inpecting that one for useful checks, good stuff! :-)
Look forward to your feedback!

So the only other missing check I could find was the file permissions check:
image

**Comment by nul800sebastiaan** [comment]
I've updated all the hardcoded strings to take values from the language files (I checked if overruling the language file using `~/config/lang/xx-yy.xml` works).

- Not sure about making a private field everywhere for `_textService`
- Not sure how to best translate attributes.. they do have a unique GUID, maybe we can amend `ILocalizedTextService` to do something with that? Dunno.

File: src/Umbraco.Web/HealthCheck/Checks/Config/AbstractConfigCheck.cs [hunk 1]
```diff
@@ -3,12 +3,14 @@
...
 using System.Linq;
 using System.Web;
+using Umbraco.Core.Services;
 
 namespace Umbraco.Web.HealthCheck.Checks.Config
 {
     public abstract class AbstractConfigCheck : HealthCheck
     {
         private readonly ConfigurationService _configurationService;
+        private readonly ILocalizedTextService _textService;
```
Comment (Shazwazza) [inline src/Umbraco.Web/HealthCheck/Checks/Config/AbstractConfigCheck.cs]: might as well make this a protected property so it can be used in sub-classes

**Comment by abjerner** [comment]
@nul800sebastiaan Sorry for not answering before now. I caught a nasty flu (or bug if you will), and was down for about a week.

In general, the Health Check dashboard now looks really nice.

But I'm missing some of the UI changes that I made earlier - eg. some separation between each status of a given health check as well as the show-loader-for-at-least-200ms-thing. I understand this may be be due to you mostly copying the changes from my last pull request rather than merging. Should I just re-add these and create another pull request?

I would also prefer the rectify button to be a link, since a button takes up a lot of space in the UI. Both changes are shown on my screenshot from earlier:

image

But I'm really just a backend developer, so perhaps you can convince your design guys to shine it up a bit, and then we can have another look at it?

I also spotted a few minor issues in the UI:

- Rectifying the *Custom Errors* check will result in the status: `Custom errors successfully set to 'Off'.` Rectifying the check will set the value to `RemoteOnly`, whereas `Off` was the old value. The other checks seems to work fine, but I'll do some more testing.

- Running a check again, the UI isn't "cleaned up". Eg. when running the check, the UI will show a single line with a text like ´Loading status... Debug compilation mode successfully disabled.`. The old rectify is still shown when the next check has completed. So perhaps the UI should clear the rectify result once running a new check.

I can look into these in the coming days ;)

**Comment by nul800sebastiaan** [comment]
@abjerner Just as a quick update (I'll reply more later this week) - I didn't bring in any of your UX on purpose, the UX guys here are working on it this week as they have ideas for that!

**Comment by nul800sebastiaan** [comment]
Thanks everybody so much for all the help, this has turned into a wonderful feature, ready for inclusion in the next version. We'll keep polishing it, but it's already really nice.
#h5yr to everyone involved, many valuable contributions!

**Comment by abjerner** [comment]
Just installed the new 7.5-beta. The new health check dashboard looks awesome :+1:

**Pull request merged**